pkg/scheduler/queue/tree_queue_test.go

```diff
@@ -10,73 +10,694 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-// TestDequeueBalancedTree checks dequeuing behavior from a balanced tree.
-//
-// Dequeuing from a balanced tree allows the test to have a simple looped structures
-// while running checks to ensure that round-robin order is respected.
-func TestDequeueBalancedTree(t *testing.T) {
-	firstDimensions := []string{"0", "1", "2"}
-	secondDimensions := []string{"a", "b", "c"}
-	itemsPerDimension := 5
-	root := makeBalancedTreeQueue(t, firstDimensions, secondDimensions, itemsPerDimension)
-	require.NotNil(t, root)
+func newTenantQuerierAssignments() *tenantQuerierAssignments {
+	return &tenantQuerierAssignments{
+		tenantQuerierIDs: make(map[TenantID]map[QuerierID]struct{}),
+		tenantNodes:      make(map[string][]*Node),
+	}
+}
 
-	count := 0
-	// tree queue will fairly dequeue from all levels of the tree
-	rotationsBeforeRepeat := len(firstDimensions) * len(secondDimensions)
-	// track dequeued paths to ensure round-robin dequeuing does not repeat before expected
-	dequeuedPathCache := make([]QueuePath, rotationsBeforeRepeat)
+func Test_NewTree(t *testing.T) {
+	tests := []struct {
+		name      string
+		treeAlgos []DequeueAlgorithm
+		state     *tenantQuerierAssignments
+		expectErr bool
+	}{
+		{
+			name:      "create round-robin tree",
+			treeAlgos: []DequeueAlgorithm{&roundRobinState{}},
+		},
+		{
+			name:      "create shuffle-shard tree",
+			treeAlgos: []DequeueAlgorithm{newTenantQuerierAssignments()},
+		},
+		{
+			name:      "create shuffle-shard tree with no tenant-querier map, we should create an empty one",
+			treeAlgos: []DequeueAlgorithm{&tenantQuerierAssignments{}},
+		},
+		{
+			name:      "fail to create tree without defined dequeuing algorithm",
+			treeAlgos: []DequeueAlgorithm{nil},
+			expectErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := NewTree(tt.treeAlgos...)
+			if tt.expectErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
 
-	for !root.IsEmpty() {
-		v := root.Dequeue()
-		dequeuedPath := getChildPathFromQueueItem(v)
+		})
+	}
+}
 
-		// require dequeued path has not repeated before the expected number of rotations
-		require.NotContains(t, dequeuedPathCache, dequeuedPath)
+func Test_EnqueueBackByPath(t *testing.T) {
```

Comment:
I am having trouble understanding what `Test_EnqueueBackByPath` and each case is really trying to test, other than the enqueue beyond max depth one.

This one is more direct:
> "enqueue shuffle-shard node with no tenant-querier map to round-robin node, we should create an empty one"

But we don't test the tree structure after to validate "we should create an empty one".

I think it's worth having more extensive tests here which reach into the tree and validate the expected structure & state after various enqueues. The tests could just be for enqueues, as the tree structure changes for dequeuing are less complicated and less impactful.


nit: the field `children` is a bit confusing, `childPathsToEnqueue` might show that it's going to be an arg / input - and the use of `{"shuffle-shard-child-1}"` and `{"round-robin-child-1"}` sort of make it seem like the name of the path is tied to the algorithm - I like the `{{"child"}, {"child, grandchild"}}` approach better

Reply:
> But we don't test the tree structure after to validate "we should create an empty one".

Some of these tests (the one you mentioned, and `"create tree with multiple shuffle-shard nodes"`) are an artifact of a previous iteration, which had more expectations of the user (e.g., requiring the user to pass `tenantQuerierAssignments` with a non-nil tenant-querier map, or explicitly disallowing >1 `tenantQuerierAssignment` layer in the tree), and used to expect an error; since the expectations are no longer in place, I don't think tests that enforce those expectations are necessary anymore, and I can axe them.

> I think it's worth having more extensive tests here which reach into the tree and validate the expected structure & state after various enqueues.

I would _really_ prefer to avoid writing tests that rely on peeping the tree structure, and would generally like to move our tests away from doing so; IMO the tree should, as much as possible, be judged based on whether it produces the results we expect under various conditions, and tests ought to reproduce the conditions we want to test behavior for.

Reply:
> I would really prefer to avoid writing tests that rely on peeping the tree structure

I am fine with this. I remembered that for better or for worse, the `isConsistent` checks really reach down into the tree structure - we can re-evaluate the right place to put these when we flip the tree maybe.

My mentality is essentially if _any_ tests should be checking the inside of the tree, it's the tree_queue tests, and the others should be just checking enqueue/dequeue behavior as a black box.
Right now we are kind of in the opposite situation, but changing too much in this PR is probably going to destroy and semblance of comparing the old tests with the new for now, so let's just leave it.